include/seqan3/core/algorithm/concept.hpp

```diff
@@ -0,0 +1,150 @@
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//     * Neither the name of Knut Reinert or the FU Berlin nor the names of
+//       its contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL KNUT REINERT OR THE FU BERLIN BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ============================================================================
+
+/*!\file
+ * \brief Provides concepts for the configuration classes.
+ * \author Rene Rahn <rene.rahn AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <functional>
+#include <type_traits>
+
+#include <meta/meta.hpp>
+
+#include <seqan3/core/metafunction/template_inspection.hpp>
+#include <seqan3/std/concept/callable.hpp>
+#include <seqan3/std/concept/core_language.hpp>
+#include <seqan3/std/concept/object.hpp>
+
+namespace seqan3::detail
+{
+
+// ----------------------------------------------------------------------------
+// Concept config_element_concept
+// ----------------------------------------------------------------------------
+
+//!\cond
+template <typename derived_t>
+class config_element_base;
+//!\endcond
+
+/*!\interface seqan3::detail::config_element_concept <>
+ * \brief Concept for an algorithm configuration.
+ * \ingroup core_algorithm
+ *
+ * \extends seqan3::default_constructible_concept
+ * \extends seqan3::assignable_concept
+ * \implements seqan3::detail::config_element_base
+ *
+ * \details
+ *
+ * Classes must extend the abstract crtp-base class seqan3::detail::config_element_base in order to satisfy this
+ * concept. This concept is merely used for internal purposes and shall not be exposed in public user interfaces.
+ */
+//!\cond
+template <typename config_t>
+concept bool config_element_concept = default_constructible_concept<std::remove_reference_t<config_t>> &&
```

Comment:
Are you sure you don't want semi-regular or regular instead?

Reply:
Not sure. Actually I want to drop the default constructible requirement to pass config states per std::ref

Reply:
Many styles recommend for library types to always be semi_regular if possible. It makes many things easier, it's the style of range-v3 and I think they want to adopt it for the core guidelines, too.